docs/source/migration.mdx

```diff
@@ -407,130 +401,143 @@ interface ContextValue {
 const server = new ApolloServer({
   typeDefs,
   resolvers,
-  context: ({ req: ExpressRequest }): Omit<ContextValue, 'dataSources'> => {
+  context: ({ req: ExpressRequest }): Omit<ContextValue, 'dataSources'> => { //highlight-line
     return {
       token: getTokenFromRequest(req),
     };
   },
+  //highlight-start
   dataSources: (): ContextValue['dataSources'] => {
     return {
       moviesAPI: new MoviesAPI(),
     };
   },
+  //highlight-end
 });
 
 await server.listen();
 ```
 
 </MultiCodeBlock>
 
-<!-- TODO(AS4): Update the next line before release -->
-
-Below is how you write the same code in Apollo Server 4. Note that the `@apollo/datasource-rest` package is now renamed `@apollo/datasource-rest`.
-
-> ⚠️ This feature is in active development, and the below code snippets **do not** currently work as described. [See above for more details](#datasources).
+Below is how you write the same code in Apollo Server 4.
 
 <MultiCodeBlock>
 
-```ts {5, 25-27, 38, 41-43}  title="Apollo Server 4"
-import { RESTDataSource, RESTDataSourceOptions } from '@apollo/datasource-rest';
+```ts title="Apollo Server 4"
+import { RESTDataSource, WillSendRequestOptions } from '@apollo/datasource-rest';
+// KeyValueCache is the type of Apollo server's default cache
+import type { KeyValueCache } from '@apollo/utils.keyvaluecache';
 import { ApolloServer } from '@apollo/server';
 import { startStandaloneServer } from '@apollo/server/standalone';
 
-class MoviesAPI extends RESTDataSource {
-  baseURL = 'https://movies-api.example.com/';
+class MoviesAPI extends RESTDataSource { // highlight-line
+  override baseURL = 'https://movies-api.example.com/';
   private token: string;
 
-  constructor(options: { token: string } & RESTDataSourceOptions) {
-    super(options);  // this should send `cache` through
+  constructor(options: { token: string; cache: KeyValueCache }) {
+    super(options); // this sends our server's `cache` through
     this.token = options.token;
   }
 
-  willSendRequest(request: RequestOptions) {
-    request.headers.set('Authorization', this.token);
+  override willSendRequest(request: WillSendRequestOptions) {
+    request.headers['authorization'] = this.token;
```

Comment:
we should probably mention somewhere that `request` now uses `@apollo/fetcher` like in `#apolloutilsfetcher-replaces-apollo-server-env` which means no `Request`/`Headers` methods

Reply:
I added a note above this code snippet:

> **New in Apollo Server 4**: Apollo Server 4 now uses [`@apollo/fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) under-the-hood, so you can no longer directly use `Request` or `Headers` methods.

Just to double-check, does this capture the point you are making above?

Reply:
I guess, though it's more like we've actually changed the API folks write their `this.post` etc with. Compare the TS declarations:
https://www.runpkg.com/?apollo-datasource-rest@3.7.0/dist/RESTDataSource.d.ts
https://www.runpkg.com/?@apollo/datasource-rest@4.2.0/dist/RESTDataSource.d.ts

Like, the old API each of these take three args (path, body, other options) whereas in the new API it's two args (path, request where the body comes as a `body` option).

Additionally for hooks like `willSendRequest`, the object you have access to doesn't have Request/Header methods. But the bigger bit I guess is that the API changed! I think you have to adjust for `body` in the POST and PUT examples like you already did for patch.

Reply:
1. I wasn't aware the args had for the HTTP convenience methods; thank you! So I'll **remove** this paragraph because it no longer applies:
> For all methods, the **third parameter** is an `init` object that enables you to provide additional options (such as headers and referrers) to the `fetch` API that's used to send the request. For details, [see MDN's fetch docs](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters).

2. I'm going to add this note to the "Intercepting Fetches" section to let folks know they can't access Request/Header methods directly:

>  **New in Apollo Server 4**: Apollo Server 4 now uses [`@apollo/fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) under-the-hood for fetching, so hooks like `willSendRequest` no longer have direct access to `Request` or `Headers` methods.

3.  Great catch on that `POST` example, thanks!

Reply:
For posterity, the package name is `@apollo/utils.fetcher`. (maybe you figured this out further down, I see this PR landed and I don't see any references to `@apollo/fetcher` 😄)

Reply:
That did get updated, thanks for calling this out, though!